aiida_mlip/data/model.py

```diff
@@ -0,0 +1,243 @@
+"""Define Model Data type in AiiDA"""
+
+import hashlib
+from pathlib import Path
+from typing import Any, Optional, Union
+from urllib import request
+from urllib.parse import urlparse
+
+from aiida.orm import SinglefileData
+
+
+def _calculate_hash(file: Union[str, Path]) -> str:
+    """Calculate the hash of a file.
+
+    Parameters
+    ----------
+    file : Union[str, Path]
+        Path to the file for which hash needs to be calculated.
+
+    Returns
+    -------
+    str
+        The SHA-256 hash of the file.
+    """
+    # Calculate hash
+    buf_size = 65536  # reading 64kB (arbitrary) at a time
+    sha256 = hashlib.sha256()
+    with open(file, "rb") as f:
+        # calculating sha in chunks rather than 1 large pass
+        while data := f.read(buf_size):
+            sha256.update(data)
+    file_hash = sha256.hexdigest()
+    return file_hash
+
+
+def _check_existing_file(
+    arch_path: Union[str, Path],
+    file: Union[str, Path],
+    cache_path: Union[str, Path],
+    architecture: Optional[str] = None,
+) -> tuple[Path, str]:
+    """Check if a file already exists and return its path and architecture.
+
+    Parameters
+    ----------
+    arch_path : Union[str, Path]
+        Path to the folder containing the file.
+    file : Union[str, Path]
+        Path to the file.
+    cache_path : Union[str, Path]
+        Path to the parent folder of arch_path.
+    architecture : Optional[str]
+        MLIP architecture of the model.
+
+    Returns
+    -------
+    tuple[Path, str]
+        A tuple containing the path of the file of interest and its architecture.
+    """
+    file_hash = _calculate_hash(file)
+
+    def is_diff_file(curr_path: Path):
+        return curr_path.is_file() and not curr_path.samefile(file)
+
+    for existing_file in filter(is_diff_file, arch_path.rglob("*")):
+        if _calculate_hash(existing_file) == file_hash:
+            ex_file_path = Path(existing_file).parent
+            if arch_path == ex_file_path:
+                file.unlink()
+                return existing_file, architecture
+
+            if arch_path == cache_path and arch_path != ex_file_path:
```

Comment:
```suggestion
            if arch_path == cache_path
```
If `arch_path == ex_file_path`, wouldn't the return above be triggered?